### PR TITLE
refactor(options): remove `.indir` and redesign option scopes 

### PR DESCRIPTION
**Pull request by famiu**
Problem: The way option scopes currently work is inflexible and does not allow for nested option scopes or easily finding the value of an option at any arbitrary scope without having to do long handwritten switch-case statements like in `get_varp()`. `.indir` is also confusing and redundant since option indices for each scope can be autogenerated.

Solution: Expand option scopes in such a way that an option can support any amount of scopes using a set of scope flags, similarly to how it's already done for option types. Also make options contain information about its index at each scope it supports. This allows for massively simplifying `get_varp()` and `get_varp_scope()` in the future by just using a struct for options at each scope. This would be done by creating a table that stores the offset of an option's variable at a scope by using the option's index at that scope as a key. This PR also autogenerates enums for option indices at each scope to remove the need for `.indir` entirely, and also to allow easily iterating over options all options that support any scope.

Ref: #29314

### CODE REVIEW

File: src/nvim/option.c [hunk 1]
```diff
@@ -3299,9 +3296,71 @@ bool is_option_hidden(OptIndex opt_idx)
...
+  // in the scope_flags bit field.
+  return get_option(opt_idx)->scope_flags & (1 << scope);
+}
+
+/// Check if option is global-local.
 static inline bool option_is_global_local(OptIndex opt_idx)
 {
-  return opt_idx == kOptInvalid ? false : (options[opt_idx].indir & PV_BOTH);
+  // Global-local options have at least two types, so their type flag cannot be a power of two.
+  return opt_idx == kOptInvalid ? false : !is_power_of_two(options[opt_idx].scope_flags);
```
Comment (zeertzjq) [inline src/nvim/option.c:3337]:
```suggestion
  return opt_idx != kOptInvalid && !is_power_of_two(options[opt_idx].scope_flags);
```

File: src/nvim/option.c [hunk 2]
```diff
@@ -3299,9 +3296,71 @@ bool is_option_hidden(OptIndex opt_idx)
...
+}
+
+/// Check if option only supports global scope.
+static inline bool option_is_global_only(OptIndex opt_idx)
+{
+  // For an option to be global-only, it has to only have a single scope, which means the scope
+  // flags must be a power of two, and it must have the global scope.
+  return opt_idx == kOptInvalid ? false
+                                : (is_power_of_two(options[opt_idx].scope_flags)
+                                   && option_has_scope(opt_idx, kOptScopeGlobal));
```
Comment (zeertzjq) [inline src/nvim/option.c:3347]: ditto

File: src/nvim/option.c [hunk 3]
```diff
@@ -3299,9 +3296,71 @@ bool is_option_hidden(OptIndex opt_idx)
...
+}
+
+/// Check if option only supports window scope.
+static inline bool option_is_window_local(OptIndex opt_idx)
+{
+  // For an option to be window-local it has to only have a single scope, which means the scope
+  // flags must be a power of two, and it must have the window scope.
+  return opt_idx == kOptInvalid ? false
+                                : (is_power_of_two(options[opt_idx].scope_flags)
+                                   && option_has_scope(opt_idx, kOptScopeWindow));
```
Comment (zeertzjq) [inline src/nvim/option.c:3357]: ditto

File: src/nvim/api/vim.c [hunk 4]
```diff
@@ -1004,10 +1004,10 @@ Buffer nvim_create_buf(Boolean listed, Boolean scratch, Error *err)
   buf_copy_options(buf, BCO_ENTER | BCO_NOHELP);
 
   if (scratch) {
-    set_option_direct_for(kOptBufhidden, STATIC_CSTR_AS_OPTVAL("hide"), OPT_LOCAL, 0, kOptReqBuf,
-                          buf);
-    set_option_direct_for(kOptBuftype, STATIC_CSTR_AS_OPTVAL("nofile"), OPT_LOCAL, 0, kOptReqBuf,
-                          buf);
+    set_option_direct_for(kOptBufhidden, STATIC_CSTR_AS_OPTVAL("hide"), OPT_LOCAL, 0,
+                          kOptScopeBuffer, buf);
```
Comment (justinmk) [inline src/nvim/api/vim.c:1008]:
```suggestion
                          kOptScopeBuf, buf);
```

File: src/nvim/api/options.c [hunk 5]
```diff
@@ -39,14 +39,14 @@ static int validate_option_value_args(Dict(option) *opts, char *name, OptIndex *
...
-  *req_scope = kOptReqGlobal;
+  *req_scope = kOptScopeGlobal;
 
   if (filetype != NULL && HAS_KEY_X(opts, filetype)) {
     *filetype = opts->filetype.data;
   }
 
   if (HAS_KEY_X(opts, win)) {
-    *req_scope = kOptReqWin;
+    *req_scope = kOptScopeWindow;
```
Comment (justinmk) [inline src/nvim/api/options.c:49]: `:help dev-naming`

```suggestion
    *req_scope = kOptScopeWin;
```

File: src/nvim/buffer_defs.h [hunk 6]
```diff
@@ -206,7 +206,7 @@ typedef struct {
   OptInt wo_winbl;
 #define w_p_winbl w_onebuf_opt.wo_winbl  // 'winblend'
 
-  LastSet wo_script_ctx[WV_COUNT];        // SCTXs for window-local options
+  LastSet wo_script_ctx[kWindowOptCount];  // SCTXs for window-local options
```
Comment (justinmk) [inline src/nvim/buffer_defs.h:209]:
```suggestion
  LastSet wo_script_ctx[kWinOptCount];  // SCTXs for window-local options
```

File: src/nvim/buffer_defs.h [hunk 7]
```diff
@@ -512,7 +512,7 @@ struct file_buffer {
   // or contents of the file being edited.
   bool b_p_initialized;                 // set when options initialized
 
-  LastSet b_p_script_ctx[BV_COUNT];     // SCTXs for buffer-local options
+  LastSet b_p_script_ctx[kBufferOptCount];  // SCTXs for buffer-local options
```
Comment (justinmk) [inline src/nvim/buffer_defs.h:515]:
```suggestion
  LastSet b_p_script_ctx[kBufOptCount];  // SCTXs for buffer-local options
```

File: src/nvim/options.lua [hunk 8]
```diff
@@ -81,6 +80,8 @@ end
 -- luacheck: ignore 621
 return {
   cstr = cstr,
+  --- @type string[]
+  valid_scopes = { 'global', 'buffer', 'window' },
```
Comment (justinmk) [inline src/nvim/options.lua:84]:
```suggestion
  valid_scopes = { 'global', 'buf', 'win' },
```

**Review by justinmk — APPROVED** [review]
This looks great

**Review by justinmk — APPROVED** [review]

**Pull request merged**